src/databases/Cube/CubeReader.C

```diff
@@ -1,616 +1,758 @@
-        errorString = "Cube reader: Number of orbitals does not match the number of orbitals given.";
+    if(tokens.size() >= 4)
+    {
+        y_size = atoi(tokens[0].c_str());
+        dy[0]  = atof(tokens[1].c_str());
+        dy[1]  = atof(tokens[2].c_str());
+        dy[2]  = atof(tokens[3].c_str());
+    }
+    else
+    {
+        errorString = "Cube reader: Error in line 5: Too few entries to determine y grid (expected 4).";
         return;
     }
-    debug4 << "Found " << norbitals << " orbitals" << endl;
-    for (int o=1;o<=norbitals; o++) {
-      orbitals.push_back( atoi( tokens[o].c_str() ) );
-    debug4 << "        "<< orbitals[o-1] << endl;
-    }
-  }else{
-    //Go back to the real ending of the header
-    file.seekg( endHeaderPos );
-  }
-  
-  //now the file pointer is at the beginning of the grid data
-  
-  //Close the file if the caller did not explicitly request to keep the file open
-  if( !keepFileOpen ){
-     file.close();
-  }
-  debug4 << "---------------------------------------" << endl;
+    if (y_size < 0)
+    {
+        y_size = -1 * y_size;
+        SetGridUnitsToBohr();
+        debug4 << "Setting grid units to bohr" << endl;
+    }
+    if ((dy[0] != 0) || (dy[2] != 0))
+        isSheared = true;
+
+    // 6. line: Read the z grid
+    getline(file, header);
+    tokens.clear();
+    iss.clear();
+    iss.str(header);
+    copy(istream_iterator<string>(iss), istream_iterator<string>(), back_inserter<vector<string> >(tokens));
+    if(tokens.size() >= 4)
+    {
+        z_size = atoi(tokens[0].c_str());
+        dz[0]  = atof(tokens[1].c_str());
+        dz[1]  = atof(tokens[2].c_str());
+        dz[2]  = atof(tokens[3].c_str());
+    }
+    else
+    {
+        errorString = "Cube reader: Error in line 6: Too few entries to determine x grid (expected 4).";
+        return;
+    }
+    if (z_size < 0)
+    {
+        z_size = -1 * z_size;
+        SetGridUnitsToBohr();
+        debug4 << "Setting grid units to bohr" << endl;
+    }
+    if ((dz[0] != 0) || (dz[1] != 0))
+        isSheared = true;
+
+    debug4 << "Grid size is " << x_size << "x"
+           << y_size << "x"
+           << z_size << endl;
+
+    debug4 << "Step size is"
+           << "(" << dx[0] << "," << dx[1] << "," << dx[2] << ") x "
+           << "(" << dy[0] << "," << dy[1] << "," << dy[2] << ") x "
+           << "(" << dz[0] << "," << dz[1] << "," << dz[2] << ")"
+           << endl;
+
+    // Line 7 to 7+natoms :
+    // Read for each atom its type, isotope and x ,y , z location
+    atom_types.clear();
+    atom_locations.clear();
+
+    for (int i = 0; i < natoms; i++)
+    {
+        //float isotope;
+        getline(file, header);
+        tokens.clear();
+        iss.clear();
+        iss.str(header);
+        copy(istream_iterator<string>(iss), istream_iterator<string>(), back_inserter<vector<string> >(tokens));
+        if(tokens.size() >= 5)
+        {
+            atom_types.push_back(atoi(tokens[0].c_str()));
```

Comment:
Again, I know it predates your changes here but `strtod()` with error checking is a better choice than `atof()`. I will file an issue ticket for this across the board in our ascii readers.